File: src/cpp/device_detection.cpp
#include "device_detection.h"
#include "sdk-sys.h"

namespace fastly::device_detection {

fastly::expected<std::optional<Device>> lookup(std::string_view user_agent) {
  fastly::sys::device_detection::Device *out;
  fastly::sys::error::FastlyError *err;
  fastly::sys::device_detection::f_device_detection_lookup(
      static_cast<std::string>(user_agent), out, err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else if (out != nullptr) {
    return std::optional<Device>(FSLY_BOX(device_detection, Device, out));
  } else {
    return std::nullopt;
  }
}

std::optional<std::string> Device::device_name() {
  std::string out;
  auto some{this->dev->device_name(out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{out};
  }
}

std::optional<std::string> Device::brand() {
  std::string out;
  auto some{this->dev->brand(out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{out};
  }
}

std::optional<std::string> Device::model() {
  std::string out;
  auto some{this->dev->model(out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{out};
  }
}

std::optional<std::string> Device::hwtype() {
  std::string out;
  auto some{this->dev->hwtype(out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{out};
  }
}

std::optional<bool> Device::is_ereader() {
  auto ptr{this->dev->is_ereader()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_gameconsole() {
  auto ptr{this->dev->is_gameconsole()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_mediaplayer() {
  auto ptr{this->dev->is_mediaplayer()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_mobile() {
  auto ptr{this->dev->is_mobile()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_smarttv() {
  auto ptr{this->dev->is_smarttv()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_tablet() {
  auto ptr{this->dev->is_tablet()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_tvplayer() {
  auto ptr{this->dev->is_tvplayer()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_desktop() {
  auto ptr{this->dev->is_desktop()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_touchscreen() {
  auto ptr{this->dev->is_touchscreen()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

} // namespace fastly::device_detection

File: src/device_detection.rs
use std::pin::Pin;

use cxx::CxxString;

use crate::{error::ErrPtr, try_fe};

pub struct Device(pub(crate) fastly::device_detection::Device);

// Just used to force generation of symbols, since we're otherwise doing raw pointers.
pub fn f_device_detection_noop(dev: Box<Device>) -> Box<Device> {
    dev
}

pub fn f_device_detection_lookup(
    user_agent: &CxxString,
    mut out: Pin<&mut *mut Device>,
    mut err: ErrPtr,
) {
    out.set(
        fastly::device_detection::lookup(try_fe!(err, user_agent.to_str()))
            .map(Device)
            .map(Box::new)
            .map(Box::into_raw)
            .unwrap_or_else(std::ptr::null_mut),
    )
}

macro_rules! maybe_props {
    ( $( $name:ident ),* ) => {
        impl Device {
            $(
                pub fn $name(&self, out: Pin<&mut CxxString>) -> bool {
                    self.0.$name().map(|val| out.push_str(val)).is_some()
                }
            )*
        }
    }
}

macro_rules! maybe_predicates {
    ( $( $name:ident ),* ) => {
        impl Device {
            $(
                pub fn $name(&self) -> *const bool {
                    if let Some(b) = self.0.$name() {
                        Box::into_raw(Box::new(b))
                    } else {
                        std::ptr::null()
                    }
                }
            )*
        }
    }
}

maybe_props![device_name, brand, model, hwtype];
maybe_predicates![
    is_ereader,
    is_gameconsole,
    is_mediaplayer,
    is_mobile,
    is_smarttv,
    is_tablet,
    is_tvplayer,
    is_desktop,
    is_touchscreen
];
